core/src/main/java/org/mule/runtime/core/internal/routing/FirstSuccessful.java

```diff
@@ -6,42 +6,131 @@
+import static org.mule.runtime.api.i18n.I18nMessageFactory.createStaticMessage;
+import static org.mule.runtime.core.api.lifecycle.LifecycleUtils.initialiseIfNeeded;
+import static org.mule.runtime.core.internal.event.EventQuickCopy.quickCopy;
+import static org.mule.runtime.core.privileged.processor.MessageProcessors.*;
+import static reactor.core.publisher.Flux.from;
 
 /**
  * FirstSuccessful routes an event to the first target route that can accept it without throwing or returning an exception. If no
  * such route can be found, an exception is thrown. Note that this works more reliable with synchronous targets, but no such
  * restriction is imposed.
  */
-public class FirstSuccessful extends AbstractOutboundRouter {
+public class FirstSuccessful extends AbstractComponent implements Router, Lifecycle, MuleContextAware {
+
+  private final List<ProcessorRoute> routes = new ArrayList<>();
+  private MuleContext muleContext;
 
-  private RoutingStrategy routingStrategy;
 
   @Override
   public void initialise() throws InitialisationException {
-    super.initialise();
-    routingStrategy =
-        new FirstSuccessfulRoutingStrategy(this::doProcessRoute);
+    for (ProcessorRoute route : routes) {
+      initialiseIfNeeded(route, muleContext);
+    }
+  }
+
+  @Override
+  public void start() throws MuleException {
+    for (ProcessorRoute route : routes) {
+      route.start();
+    }
   }
 
-  /**
-   * Route the given event to one of our targets
-   */
   @Override
-  public CoreEvent route(CoreEvent event) throws MuleException {
-    try {
-      return routingStrategy.route(event, getRoutes());
-    } catch (RoutingFailedException e) {
-      throw new CouldNotRouteOutboundMessageException(this, e);
+  public void stop() throws MuleException {
+    for (ProcessorRoute route : routes) {
+      route.stop();
     }
   }
 
   @Override
-  public boolean isMatch(CoreEvent event, CoreEvent.Builder builder) throws MuleException {
-    return true;
+  public void dispose() {
+    for (ProcessorRoute route : routes) {
+      route.dispose();
+    }
+  }
+
+  @Override
+  public void setMuleContext(MuleContext context) {
+    this.muleContext = context;
   }
+
+  @Override
+  public CoreEvent process(CoreEvent event) throws MuleException {
+    return processToApply(event, this);
+  }
+
+  public void addRoute(final Processor processor) {
+    routes.add(new ProcessorRoute(processor));
+  }
+
+  public void setRoutes(Collection<Processor> routes) {
+    routes.forEach(this::addRoute);
+  }
+
+  @Override
+  public Publisher<CoreEvent> apply(Publisher<CoreEvent> publisher) {
+    return from(publisher)
+        .flatMap(event -> Mono.from(processWithChildContext(event, new FirstSuccessfulRoutingProcessor(), of(getLocation()))));
+  }
+
+  private class FirstSuccessfulRoutingProcessor implements Processor {
+
+    @Override
+    public CoreEvent process(CoreEvent event) throws MuleException {
+      CoreEvent returnEvent = null;
```

Comment:
This is the same logic that was in the FSRoutingStrategy, I simply transformed it in a Processor